docs/design/apiserver_storage.md

```diff
@@ -0,0 +1,153 @@
+
+The solution we proposed here is to start a storage layer from scratch. As Daniel Smith [mentioned](https://github.com/kubernetes/kubernetes/pull/28508#issuecomment-232508836)
+
+> Just to (probably unnecessarily) reiterate my position, I think we're going to need side-by-side implementations of both old and new storage styles, so we shouldn't end up locked into whatever we do now.
+
+We also believe we should start a side-by-side storage implementation to reach the goal we listed. That way, we won’t be affected by any existing issues, and it gives us the opportunities to rethink the design from the ground up, to achieve the goals more clearly. The only risk is that the effort will be completely wasted, if we cannot reach the goals.
+
+## Proposed Storage API
+
+```go
+
+type Storage interface {
+	// Wait waits until the storage has seen any version >= given version.
+	Wait(ctx context.Context, version uint64)
+
+	// Put puts an object with a key based on the given conditions.
+	Put(ctx context.Context, key string, obj runtime.Object, conditions *Conditions) (cur runtime.Object, err error)
+
+	// Delete a key and its object based on the given conditions.
+	Delete(ctx context.Context, key string, conditions *Conditions) (cur runtime.Object, err error)
+
+	// Get gets the most recent version of a key.
+	// A key that was modified by this interface will be reflected on read after write succeeded.
+	// If no object exists on the key, it will return not found error.
+	Get(ctx context.Context, key string) (cur runtime.Object, err error)
+
+	// List lists all objects that has given prefix and satisfies selectors.
+	List(ctx context.Context, prefix string, ss ...Selector) (objects []runtime.Object, globalRev uint64, err error)
+
+	// WatchPrefix watches a prefix after given version. If version is 0, we will watch from current state.
+	// It returns notifications of any keys that has given prefix.
+	// Given selectors, it returns events that contained object of interest, either of current and previous.
+	// If there is any problem establishing the watch channel, it will return error.
+	// After channel is established, any error that happened will be returned from WatchChan
+	// immediately before it's closed.
+	WatchPrefix(ctx context.Context, prefix string, version uint64, ss ...Selector) (WatchChan, error)
+
+	// AddIndex adds a new object index.
+	AddIndex(indexName, field string, g FieldValueGetFunc) error
+	// DeleteIndex deletes an object index.
+	DeleteIndex(indexName string)
+
+	// AddWatcherIndex adds a new watcher index.
+	AddWatcherIndex(indexName, field string, g FieldValueGetFunc) error
```

Comment:
do we need the ability for multi-valued indices?
/cc @derekwaynecarr for index input